stylegan_code_finder/train_with_lightning.py

```diff
@@ -69,45 +51,39 @@ def main(rank: int, args: argparse.Namespace, world_size: int):
     else:
         val_data_loader = None
 
-    train_builder_class = get_train_builder_class(config)
-    training_builder = train_builder_class(config, train_data_loader, val_data_loader, rank=rank, world_size=world_size)
-
-    pathlib.Path(args.log_dir, args.log_name).mkdir(parents=True, exist_ok=True)
+    config['num_iter_epoch'] = len(train_data_loader)
+    Path(args.log_dir, args.log_name).mkdir(parents=True, exist_ok=True)
 
     logging.info("Initializing wandb... ")
     logger = WandbLogger(name=args.log_name, save_dir=args.log_dir, project=args.wandb_project_name)
     logging.info("done")
 
-    plot_data_loader = val_data_loader
-    if val_data_loader is None:
-        plot_data_loader = train_data_loader
-    segmentation_plotter = SegmentationPlotter(plot_data_loader, config)
     segmenter_class = get_segmenter_class(config)
-    segmenter = segmenter_class(training_builder, config, segmentation_plotter, logger)
-
-    #snapshotter = training_builder.get_snapshotter()
-    #if snapshotter is not None:
-    #    trainer.extend(snapshotter)
+    segmenter = segmenter_class(config)
 
-    #image_plotter = training_builder.get_image_plotter()
-    #if image_plotter is not None:
-    #    trainer.extend(image_plotter)
-
-    if 'max_iter' in config:
-        config['epochs'] = 1000
+    if config['network'] == 'EMANet':
+        multiprocessing_strategy = 'ddp'
     else:
-        config['max_iter'] = -1
+        multiprocessing_strategy = 'ddp_find_unused_parameters_false'
+
+    model_checkpoint_callback = ModelCheckpoint(save_top_k=10, monitor='val_loss',
+                                                dirpath=Path(config['log_dir'], 'checkpoints'),
+                                                filename="segmentation-{epoch:02d}-{val_loss:.2f}")
+    callbacks = [EarlyStopping(monitor='val_loss', patience=config['patience']), model_checkpoint_callback]
 
-    synchronize()
     logging.info('Setup complete. Starting training...')
     try:
-        segmentation_trainer = pl.Trainer(logger=logger, max_epochs=config['epochs'], max_steps=config['max_iter'],
-                                          accelerator='gpu', devices=world_size)
-        segmentation_trainer.fit(segmenter, train_data_loader, val_data_loader) #ckpt_path = None
+        if 'max_iter' in config:
+            segmentation_trainer = pl.Trainer(strategy=multiprocessing_strategy, logger=logger,
+                                              log_every_n_steps=config['log_iter'], max_steps=config['max_iter'],
+                                              accelerator='gpu', devices='auto', callbacks=callbacks)
+        else:
+            segmentation_trainer = pl.Trainer(strategy=multiprocessing_strategy, logger=logger,
+                                              log_every_n_steps=config['log_iter'], max_epochs=config['epochs'],
+                                              accelerator='gpu', devices='auto', callbacks=callbacks)
```

Comment:
Lots of these parameters are the same, which makes it a bit hard to directly see where the difference is. Maybe one could use `functools.partial` here to create a partial in the if statements and then initialize the rest of the parameters.  
However, it might just be the best to use an if else statement to initialize two variables (maybe `max_epochs` and `max_epochs`) to the needed value and the default value (probably -1) respectively. Afterwards call the trainer once and pass both variables